components/pluralrules/src/rules/ast.rs

```diff
@@ -0,0 +1,436 @@
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+///
+/// Expression {
+///     operand: Operand::I,
+///     modulus: Some(Value(100)),
+/// };
+///
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Expression {
+    pub operand: Operand,
+    pub modulus: Option<Value>,
+}
+
+/// An incomplete AST representation of a plural rule. Comprises a char.
+///
+/// # Examples
+///
+/// All AST nodes can be built explicitly, as seen in the example. However, due to its complexity, it is preferred to build the AST using the parse_plural_rule function.
+///
+/// ```text
+/// "i"
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::Operand;
+///
+/// Operand::I;
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub enum Operand {
+    /// Absolute value of input
+    N,
+    /// Integer value of input
+    I,
+    /// Number of visible fraction digits with trailing zeros
+    V,
+    /// Number of visible fraction digits without trailing zeros
+    W,
+    /// Visible fraction digits with trailing zeros
+    F,
+    /// Visible fraction digits without trailing zeros
+    T,
+}
+
+/// An incomplete AST representation of a plural rule. Comprises a vector of RangeListItems.
+///
+/// # Examples
+///
+/// All AST nodes can be built explicitly, as seen in the example. However, due to its complexity, it is preferred to build the AST using the parse_plural_rule function.
+///
+/// ```text
+/// "5, 7, 9"
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+///
+/// RangeList(Box::new([
+///     RangeListItem::Value(Value(5)),
+///     RangeListItem::Value(Value(7)),
+///     RangeListItem::Value(Value(9)),
+/// ]));
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct RangeList(pub Box<[RangeListItem]>);
+
+/// An enum of items that appear in a RangeList: Range or a Value.
+///
+/// See Range and Value for additional details.
+/// A range comprises two Values: an inclusive lower and upper limit.
+///
+/// # Examples
+///
+/// ```text
+/// 5
+/// 11..15
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+///
+/// let _ = RangeListItem::Value(Value(5));
+/// let _ = RangeListItem::Range(Value(11)..=Value(15));
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub enum RangeListItem {
+    Range(RangeInclusive<Value>),
+    Value(Value),
+}
+
+/// An incomplete AST representation of a plural rule, representing one integer.
+///
+/// # Examples
+///
+/// All AST nodes can be built explicitly, as seen in the example. However, due to its complexity, it is preferred to build the AST using the parse_plural_rule function.
+///
+/// ```text
+/// "99"
+/// ```
+///
+/// Can be represented by the AST:
+///
+/// ```
+/// use icu_pluralrules::rules::ast::*;
+///
+/// RangeListItem::Value(Value(99));
+/// ```
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+pub struct Value(pub u64);
```

Comment:
Hmm.  Would it be a good idea to do all PluralRules resolution in FixedDecimal space, instead of going back and forth to machine integer space?

If you start with an integer as input, then we would need to convert the integer to a string, which is kinda annoying.

However, if you start with a FixedDecimal as input (from either NumberFormat, a string, or a properly resolved f64), I wonder if it would be better and more efficient to implement `Eq` and `Ord` on FixedDecimal, and change `Value` to be a FixedDecimal, instead of taking the pieces of the input and converting them into machine integers to match the type of `Value`.

Reply:
PluralRules AST uses only integers It has floats only in Samples, and those are ignored at runtime.

I think the use of integers in AST and resolver is going to be much faster than converting all integers into `FixedDecimals`. Do you think otherwise?

Reply:
When you read a number from the rules string, you would read it directly into a FixedDecimal.  No math necessary.

So, for instance, if you are trying to select based on a FixedDecimal, you can compare your FixedDecimal to the one that came from the rules string.  You never venture into machine integer space.

I don't know if this is a good idea; it was just something I noticed and thought I'd point out.

Reply:
I think it's a good time to have those kind of brainstorms!

My assumption is that AST being a tree of primitive values is good for AST serialization, and memory layout, and I can't imagine `Eq<u64> for FixedDecimal` being slower than `Eq<FixedDecimal> for FixedDecimal`.
I'd like to keep it as `u64` for now and we can experiment with this value once we land this.

Is that ok with you?

Reply:
With `Eq<FixedDecimal> for FixedDecimal` you can do a straight memory comparison, as long as the digits are on the stack. It should be very fast. `Eq<u64> for FixedDecimal` means you need to either convert the u64 to a FixedDecimal or the FixedDecimal to a u64, which requires doing math.  However, the math should be very fast.

But yes, I don't feel strongly about this.  I think `u64` carries other advantages like being smaller on the stack and being fast when doing integer rule selection rather than fixed-point rule selection.

Reply:
Thanks for the explanation!

I'm going to land as-is, and we can explore switching after between now and any upcoming releases as we integrate FixedDecimal.